include/spl-lockdbg.h

```diff
@@ -0,0 +1,297 @@
+{
+	struct rb_node *n = root->rb_node;
+	lock_tracking_node_t *mtp;
+	int cmp;
+
+	while (n) {
+		mtp = rb_entry(n, lock_tracking_node_t, lt_rbnode);
+
+		if (cmp_fn == slt_ptr_cmp)
+			cmp = cmp_fn(val, (void *)mtp->lt_ptr);
+		else
+			cmp = cmp_fn(val, (void *)mtp->lt_loc);
+
+		if (cmp < 0)
+			n = n->rb_left;
+		else if (cmp > 0)
+			n = n->rb_right;
+		else
+			return (mtp);
+	}
+	return (NULL);
+}
+
+static inline void
+slt_insert(struct rb_root *root, slt_cmp_fn cmp_fn,
+    lock_tracking_node_t *mtp)
+{
+	struct rb_node **p = &(root->rb_node);
+	struct rb_node *parent = NULL;
+	lock_tracking_node_t * mtp_tmp;
+	int cmp;
+
+	while (*p) {
+		mtp_tmp = rb_entry(*p, lock_tracking_node_t, lt_rbnode);
+		parent = *p;
+
+		if (cmp_fn == slt_ptr_cmp)
+			cmp = cmp_fn((void *)(mtp->lt_ptr), (void *)(mtp_tmp->lt_ptr));
+		else
+			cmp = cmp_fn((void *)(mtp->lt_loc), (void *)(mtp_tmp->lt_loc));
+
+		if (cmp < 0)
+			p = &((*p)->rb_left);
+		else if (cmp > 0)
+			p = &((*p)->rb_right);
+		else
+			return;
+	}
+
+	rb_link_node(&mtp->lt_rbnode, parent, p);
+	rb_insert_color(&mtp->lt_rbnode, root);
+}
+
+static inline void
+slt_remove(struct rb_root *root, lock_tracking_node_t *mtp)
+{
+	rb_erase(&mtp->lt_rbnode, root);
+}
+
+static void
+spl_lock_tracking_report(spl_lock_tracking_t *sltp)
+{
+	lock_tracking_node_t *mtp;
+	struct rb_node *n;
+	struct rb_root *root;
+	ulong_t flags;
+	int i;
+
+	/* Print 'header' for mismatched calls */
+	if (SLT_MISMATCHED_NUM(sltp) > 0)
+		printk(KERN_WARNING
+		    "\n******************************************************\n"
+		    "SPL Lock Tracker [%s]: missing destroy()\n"
+		    "******************************************************\n",
+		    SLT_MODULE(sltp));
+
+	root = SLT_MISMATCHED(sltp);
+	while ((n = rb_first(root)) != NULL) {
+		mtp = rb_entry(n, lock_tracking_node_t, lt_rbnode);
+		rb_erase(&mtp->lt_rbnode, root);
+
+		/* Report mismatch */
+		printk(KERN_WARNING "[0x%p] init location = %s",
+		    (void *)mtp->lt_ptr, mtp->lt_loc);
+
+		kfree(mtp);
+	}
+
+	/* Print 'header' if lock tracking is not empty */
+	if (SLT_TRACKED_NUM(sltp) > 0)
+		printk(KERN_WARNING
+		    "\n******************************************************\n"
+		    "SPL Lock Tracker [%s]: missing teardown destroy()\n"
+		    "******************************************************\n",
+		    SLT_MODULE(sltp));
+
+	for (i = 0; i < SLT_BUCKET_SIZE; i++) {
+		spin_lock_irqsave(SLT_BUCKET_LOCK(sltp, i), flags);
+		root = SLT_BUCKET(sltp, i);
+
+		while ((n = rb_first(root)) != NULL) {
+			mtp = rb_entry(n, lock_tracking_node_t, lt_rbnode);
+			rb_erase(&mtp->lt_rbnode, root);
+
+			/* Report mismatch */
+			printk(KERN_WARNING "[0x%p] init location = %s",
+			    (void *)mtp->lt_ptr, mtp->lt_loc);
+
+			kfree(mtp);
+		}
+		spin_unlock_irqrestore(SLT_BUCKET_LOCK(sltp, i), flags);
+	}
+
+
+}
+
+static void
+spl_lock_tracking_record_init(spl_lock_tracking_t *sltp, ulong_t lp, char *loc)
```

Comment:
Rather than implement everything in this header lets move all of these functions in to `module/spl/spl-lockdbg.c`.  Then we can do the conventional thing and provide structures, defines, and public prototypes in the header.